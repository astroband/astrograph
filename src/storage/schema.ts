// Unfortunately, DGraph schema doesn't support comments,
// so we'll remove them later before exporting, using regex
const schemaWithComments = `
  # common
  order: int @index(int) .
  key: string @index(hash) .
  kind.setOption: string .
  kind.payment: string .
  kind.pathPayment: string .
  kind.manageOffer: string .
  kind.manageDatum: string .
  kind.createAccount: string .
  kind.changeTrust: string .
  kind.allowTrust: string .

  # ledgers
  ledger.id: int @index(int) .
  ledger.prev: uid @reverse .
  version: int .
  base_fee: int .
  base_reserve: int .
  max_tx_set_size: int .
  close_time: datetime @index(hour) .
  _ingested: bool .

  # accounts
  account.id: string @index(exact) .
  deleted: bool @index(bool) .

  # assets
  asset.id: string @index(exact) .
  asset.issuer: uid @reverse .
  code: string @index(exact) .
  native: bool @index(bool) .

  # transactions
  tx.id: string @index(exact) .
  tx.ledger: uid @reverse .
  tx.source: uid @reverse .
  tx.index: int .
  tx.prev: uid @reverse .
  result_code: int .
  fee_amount: int .
  fee_charged: int .
  success: bool @index(bool) .
  time_bounds.min: dateTime .
  time_bounds.max: dateTime .
  memo.type: string .
  memo.value: string .

  # ledger change
  balance: string .

  # operations
  op.id: string @index(exact) .
  op.transaction: uid @reverse .
  op.ledger: uid @reverse .
  op.index: int .
  op.source: uid @reverse .
  op.destination: uid @reverse .
  op.kind: string @index(hash) .
  op.prev: uid @reverse .
  amount: int .

  # account merge
  result: uid .
  account_merge_op.result_code: int .

  # allow trust
  allow_trust_op.trustor: uid @reverse .
  allow_trust_op.result_code: int .
  allow_trust_op.asset: uid @reverse .
  authorize: bool @index(bool) .

  # bump sequence
  bump_sequence_op.result_code: int .
  bump_to: int @index(int) .

  # change trust
  change_trust_op.asset: uid @reverse .
  change_trust_op.result_code: int .
  limit: int @index(int) .

  # create account
  starting_balance: int @index(int) .
  create_account_op.result_code: int .

  # manage data
  manage_data_op.result_code: int .
  name: string @index(exact) .
  value: string @index(exact) . 

  # manage offer
  manage_offer_op.asset_buying: uid @reverse .
  manage_offer_op.asset_selling: uid @reverse .
  manage_offer_op.result_code: int .
  price_n: int .
  price_d: int .
  price: float .
  offer_id: int @index(int) .

  # payment
  payment_op.asset: uid @reverse .
  payment_op.result_code: int .

  # path payment
  path_payment_op.asset_source: uid @reverse .
  path_payment_op.asset_destination: uid @reverse .
  path_payment_op.assets_path: uid @reverse .
  path_payment_op.result_code: int .
  send_max: int .
  
  # set options
  set_options_op.signer: uid @reverse .
  set_options_op.inflation_destination: uid @reverse .
  set_options_op.result_code: int .
  set_options_op.signer: uid .
  clear_flags: int .
  set_flags: int .
  master_weight: int .
  home_domain: string .
  thresholds: uid .

  # thresholds
  high: int .
  med: int .
  low: int .

  # signer
  account: uid .
  weight: int .

  # operation results
  # account merge result
  source_account_balance: int .

  # path payment result
  path_payment_result.no_issuer: uid .
  path_payment_result.last: uid .
  destination: uid .
  last.asset: uid .
  path_payment_result.offers: uid .
  asset_sold: uid .
  asset_bought: uid .
  amount_sold: int .
  amount_bought: int .
  seller: uid .
`;

// remove comments
const validDgraphSchema = schemaWithComments.replace(/#.+$/gm, "");

export { validDgraphSchema as SCHEMA };
