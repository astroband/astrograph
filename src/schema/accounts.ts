import { gql } from "apollo-server";

export const typeDefs = gql`
  scalar AccountID

  type AccountFlags {
    authRequired: Boolean!
    authRevokable: Boolean!
    authImmutable: Boolean!
  }

  type AccountThresholds {
    masterWeight: Int!
    low: Int!
    medium: Int!
    high: Int!
  }

  interface IAccount {
    id: AccountID!
    sequenceNumber: String!
    numSubentries: Int!
    inflationDest: AccountID
    homeDomain: String
    thresholds: AccountThresholds!
    flags: AccountFlags!
    signers: [Signer]
  }

  type Account implements IAccount {
    id: AccountID!
    sequenceNumber: String!
    numSubentries: Int!
    inflationDest: AccountID
    homeDomain: String
    thresholds: AccountThresholds!
    flags: AccountFlags!
    ledger: Ledger!
    signers: [Signer]
    data: [DataEntry]
    trustLines: [TrustLine]
    signerFor(first: Int!): [Account!]
    operations(first: Int, after: String, last: Int, before: String): OperationConnection
  }

  type OperationConnection {
    pageInfo: PageInfo!
    nodes: [Operation]
    edges: [OperationEdge]
  }

  type OperationEdge {
    cursor: String!
    node: Operation
  }

  type AccountValues implements IAccount {
    id: AccountID!
    sequenceNumber: String!
    numSubentries: Int!
    inflationDest: AccountID
    homeDomain: String
    thresholds: AccountThresholds!
    flags: AccountFlags!
    signers: [Signer]
  }

  type AccountSubscriptionPayload {
    id: AccountID!
    mutationType: MutationType!
    values: AccountValues
  }

  type Signer {
    account: Account!
    signer: Account!
    weight: Int!
  }

  extend type Query {
    account(id: AccountID!): Account
    accounts(id: [AccountID!]!): [Account]
    accountsSignedBy(id: AccountID!, first: Int!): [Account!]
    signers(id: AccountID!): [Signer]
  }

  extend type Subscription {
    account(args: EventInput): AccountSubscriptionPayload
  }

`;
