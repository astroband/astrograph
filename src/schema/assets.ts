import { gql } from "apollo-server";

export const typeDefs = gql`
  "Stellar asset alphanumeric code"
  scalar AssetCode
  "String in \`{id}-{issuer}'\` format, which uniquely identifies single asset"
  scalar AssetID

  "Represents single [asset](https://www.stellar.org/developers/guides/concepts/assets.html) on Stellar network"
  type Asset {
    "Whether this asset is [lumens](https://www.stellar.org/developers/guides/concepts/assets.html)"
    native: Boolean!
    "Asset issuer's account. It's \`null\` for native lumens"
    issuer: Account
    "Asset's code"
    code: AssetCode!
    "Sum of all asset holders balances"
    totalSupply: String!
    "Sum of only authorized holders balances"
    circulatingSupply: String!
    "Total number of asset holders"
    holdersCount: Int!
    "Total number of unathorized holders"
    unauthorizedHoldersCount: Int!
    "Ledger this asset was last time modified in"
    lastModifiedIn: Ledger!
    "Requires the issuing account to give other accounts permission before they can hold the issuing accountâ€™s credit"
    authRequired: Boolean!
    "Allows the issuing account to revoke its credit held by other accounts"
    authRevocable: Boolean!
    "If this is set then none of the authorization flags can be set and the account can never be deleted"
    authImmutable: Boolean!
    "All accounts that trust this asset, ordered by balance"
    balances(first: Int, last: Int, after: String, before: String): BalanceConnection
  }

  "A list of assets"
  type AssetConnection {
    pageInfo: PageInfo!
    nodes: [Asset]
    edges: [AssetEdge]
  }

  type AssetEdge {
    cursor: String!
    node: Asset
  }

  input AssetInput {
    code: AssetCode
    issuer: AccountID
  }

  type Query {
    "Get single asset"
    asset(id: AssetID): Asset
    "Get list of assets. Note: native XLM asset isn't included here"
    assets(code: AssetCode, issuer: AccountID, first: Int, after: String, last: Int, before: String): AssetConnection
  }
`;
