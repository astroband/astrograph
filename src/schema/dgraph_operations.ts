import { gql } from "apollo-server";

export const typeDefs = gql`
  interface IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
  }

  type PaymentOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    destination: AccountID!
    asset: Asset!
    amount: String!
    source: AccountID!
  }

  type SetOptionsOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    clearFlags: Int
    setFlags: Int
    homeDomain: String
    masterWeight: Int
    thresholds: SetOptionsThresholds
    signer: SetOptionsSigner
    inflationDestination: AccountID
  }

  type AccountMergeOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    destination: AccountID!
  }

  type AllowTrustOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    trustor: AccountID
    authorize: Boolean
    assetCode: AssetCode
  }

  type BumpSequenceOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    bumpTo: Int!
  }

  type ChangeTrustOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    limit: String!
    asset: Asset!
  }

  type CreateAccountOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    startingBalance: String!
    destination: AccountID!
  }

  type ManageDatumOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    name: String!
    value: String
  }

  type ManageOfferOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    priceComponents: OfferPriceComponents!
    price: String!
    offerId: String!
    amount: String!
    assetSelling: Asset!
    assetBuying: Asset!
  }

  type PathPaymentOperation implements IOperation {
    kind: OperationKind!
    account: AccountID!
    transaction: Transaction!
    index: Int!
    dateTime: DateTime!
    sendMax: String!
    destinationAmount: String!
    destinationAsset: Asset!
    sourceAsset: Asset!
    destinationAccount: AccountID!
    sourceAccount: AccountID!
    path: [Asset]
  }

  type OfferPriceComponents {
    n: Int!
    d: Int!
  }

  type SetOptionsThresholds {
    low: Int
    medium: Int
    high: Int
  }

  type SetOptionsSigner {
    account: AccountID
    weight: Int
  }

  input PaymentOpFilter {
    destination: AccountID
    source: AccountID
    asset: AssetInput
  }

  input SetOptionsOpFilter {
    masterWeight: Int
  }

  input AccountMergeOpFilter {
    destination: AccountID
  }

  input AllowTrustOpFilter {
    trustor: AccountID
    assetCode: AssetCode
    authorize: Boolean
  }

  input BumpSequenceOpFilter {
    bumpTo: Int
  }

  input ChangeTrustOpFilter {
    limit: String
    asset: AssetInput
  }

  input CreateAccountOpFilter {
    destination: AccountID
  }

  input ManageDataOpFilter {
    name: String
    value: String
  }

  input ManageOfferOpFilter {
    assetSelling: AssetInput
    assetBuying: AssetInput
    offerId: String
  }

  input PathPaymentOpFilter {
    sourceAccount: AccountID
    destinationAccount: AccountID
    destinationAsset: AssetInput
    sourceAsset: AssetInput
    pathContains: AssetInput
  }

  input OperationsFilter {
    payment: PaymentOpFilter
    setOption: SetOptionsOpFilter
    accountMerge: AccountMergeOpFilter
    allowTrust: AllowTrustOpFilter
    bumpSequence: BumpSequenceOpFilter
    changeTrust: ChangeTrustOpFilter
    createAccount: CreateAccountOpFilter
    manageDatum: ManageDataOpFilter
    manageOffer: ManageOfferOpFilter
    pathPayment: PathPaymentOpFilter
  }

  type Query {
    dgraphAccountOperations(
      account: AccountID!
      kinds: [OperationKind]
      filters: OperationsFilter
      first: Int!
      offset: Int
    ): [IOperation]
    assetOperations(asset: AssetInput!, kinds: [OperationKind], first: Int!, offset: Int): [IOperation]
    accountTransactions(id: AccountID!, first: Int!, offset: Int): [Transaction]
  }

`;
