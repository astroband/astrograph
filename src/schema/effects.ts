import { gql } from "apollo-server";

export const typeDefs = gql`
  enum EffectKind {
    accountCreated
    accountRemoved
    accountCredited
    accountDebited
    accountThresholdsUpdated
    accountHomeDomainUpdated
    accountFlagsUpdated
    accountInflationDestinationUpdated
    signerCreated
    signerRemoved
    signerUpdated
    trustlineCreated
    trustlineRemoved
    trustlineUpdated
    trustlineAuthorized
    trustlineDeauthorized
    offerCreated
    offerRemoved
    offerUpdated
    trade
    dataCreated
    dataRemoved
    dataUpdated
    sequenceBumped
  }

  "Common [effect](https://www.stellar.org/developers/horizon/reference/resources/effect.html) attributes"
  interface Effect {
    id: String!
    "Account that this effect changed"
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type AccountCreatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    "The XLM balance new account was created with"
    startingBalance: String!
  }

  type AccountRemovedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type AccountCreditedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    amount: String!
  }

  type AccountDebitedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    amount: String!
  }

  type AccountThresholdsUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    lowThreshold: Int!
    medThreshold: Int!
    highThreshold: Int!
  }

  type AccountHomeDomainUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    homeDomain: String!
  }

  type AccountFlagsUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    authRequiredFlag: Boolean
    authRevocableFlag: Boolean
  }

  type AccountInflationDestinationUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type SequenceBumpedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    newSeq: Int!
  }

  type SignerCreatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    weight: Int!
    publicKey: AccountID!
    key: AccountID!
  }

  type SignerRemovedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    weight: Int!
    publicKey: AccountID!
    key: AccountID!
  }

  type SignerUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    weight: Int!
    publicKey: AccountID!
    key: AccountID!
  }

  type TrustlineCreatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    asset: Asset!
    limit: String!
  }

  type TrustlineRemovedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    asset: Asset!
    limit: String!
  }

  type TrustlineUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    asset: Asset!
    limit: String!
  }

  type TrustlineAuthorizedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    trustor: AccountID!
    asset: Asset!
  }

  type TrustlineDeauthorizedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    trustor: AccountID!
    asset: Asset!
  }

  type OfferCreatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type OfferRemovedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type OfferUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type TradeEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
    seller: AccountID!
    offerId: OfferID!
    soldAmount: String!
    soldAsset: Asset!
    boughtAmount: String!
    boughtAsset: Asset!
  }

  type DataCreatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type DataUpdatedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type DataRemovedEffect implements Effect {
    id: String!
    account: Account!
    kind: EffectKind!
    createdAt: DateTime!
  }

  type EffectConnection {
    pageInfo: PageInfo!
    nodes: [Effect]
    edges: [EffectEdge]
  }

  type EffectEdge {
    cursor: String!
    node: Effect
  }

  extend type Query {
    "Get list of the effects"
    effects(first: Int, last: Int, after: String, before: String): EffectConnection
  }
`;
