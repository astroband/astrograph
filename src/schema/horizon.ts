import { gql } from "apollo-server";

export const typeDefs = gql`
  interface Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
  }

  type PaymentOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    destination: AccountID!
    asset: Asset!
    amount: String!
    source: AccountID!
  }

  type SetOptionsOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    clearFlags: Int
    setFlags: Int
    homeDomain: String
    masterWeight: Int
    thresholds: SetOptionsThresholds
    signer: SetOptionsSigner
    inflationDestination: AccountID
  }

  type AccountMergeOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    destination: AccountID!
  }

  type AllowTrustOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    trustor: AccountID
    authorize: Boolean
    assetCode: AssetCode
  }

  type BumpSequenceOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    bumpTo: Int!
  }

  type ChangeTrustOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    limit: String!
    asset: Asset!
  }

  type CreateAccountOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    startingBalance: String!
    destination: AccountID!
  }

  type ManageDatumOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    name: String!
    value: String
  }

  type ManageOfferOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    priceComponents: OfferPriceComponents!
    price: String!
    offerId: String!
    amount: String!
    assetSelling: Asset!
    assetBuying: Asset!
  }

  type PathPaymentOperation implements Operation {
    id: String!
    kind: OperationKind!
    account: AccountID!
    dateTime: DateTime!
    sendMax: String!
    amountSent: String!
    amountReceived: String!
    destinationAsset: Asset!
    sourceAsset: Asset!
    destinationAccount: AccountID!
    sourceAccount: AccountID!
  }

  type OfferPriceComponents {
    n: Int!
    d: Int!
  }

  type SetOptionsThresholds {
    low: Int
    medium: Int
    high: Int
  }

  type SetOptionsSigner {
    account: AccountID
    weight: Int
  }

`;
