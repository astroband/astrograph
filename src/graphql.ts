import "./util/memo";

import * as Sentry from "@sentry/node";

import { ApolloError, ApolloServer } from "apollo-server";
import { GraphQLError } from "graphql";

import { Cursor, Worker } from "./ingest";
import schema from "./schema";
import { Connection } from "./storage/connection";
import logger from "./util/logger";
import { BIND_ADDRESS, DEBUG_LEDGER, DGRAPH_QUERY_URL, INGEST_INTERVAL, PORT, SENTRY_DSN } from "./util/secrets";
import { setNetwork as setStellarNetwork } from "./util/stellar";

if (SENTRY_DSN) {
  Sentry.init({
    dsn: SENTRY_DSN,
    integrations: [new Sentry.Integrations.RewriteFrames({ root: __dirname || process.cwd() })]
  });
}

const server = new ApolloServer({
  schema,
  tracing: true,
  introspection: true,
  playground: true,
  debug: true,
  cors: true,
  formatError: (error: ApolloError) => {
    logger.error(error);

    if (!error.originalError || error.originalError.constructor.name !== "UserInputError") {
      Sentry.captureException(error);
    }

    return new GraphQLError(
      error.message,
      error.nodes,
      error.source,
      error.positions,
      error.path,
      error,
      error.extensions
    );
  }
});

const network = setStellarNetwork();
logger.info(`Using ${network}`);

if (DGRAPH_QUERY_URL) {
  logger.info(`[DGraph] Updating schema...`);
  new Connection().migrate().catch((err: any) => {
    logger.error(err);
    Sentry.captureException(err);
    process.exit(-1);
  });
}

Cursor.build(DEBUG_LEDGER).then(cursor => {
  logger.info(
    `Staring ingest every ${INGEST_INTERVAL} ms. from ${
      DEBUG_LEDGER === -1 ? "first ledger" : DEBUG_LEDGER || "lastest ledger"
    }`
  );

  const tick = async () => {
    const worker = new Worker(cursor);
    worker
      .run()
      .then(done => {
        logger.info(`Ledger ${cursor.current}: done.`);
        if (done) {
          tick();
        } else {
          setTimeout(tick, INGEST_INTERVAL);
        }
      })
      .catch(e => {
        logger.error(e);
        if (e.message.includes("Please retry again, server is not ready to accept requests")) {
          setTimeout(tick, 200);
          return;
        }

        Sentry.captureException(e);
      });
  };

  tick();
});

server.listen({ port: PORT, host: BIND_ADDRESS }).then(({ url }) => {
  logger.info(`ğŸš€ Server ready at ${url}`);
});
