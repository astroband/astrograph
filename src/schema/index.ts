import { makeExecutableSchema, mergeSchemas } from "apollo-server";

import { typeDefs as accountsTypeDefs } from "./accounts";
import { typeDefs as horizonTypeDefs } from "./horizon";
import resolvers from "./resolvers";
import { typeDefs as transactionsTypeDefs } from "./transactions";
import { typeDefs } from "./type_defs";

const schema = makeExecutableSchema({
  typeDefs: [typeDefs, horizonTypeDefs, accountsTypeDefs, transactionsTypeDefs],
  resolverValidationOptions: { requireResolversForResolveType: false }
});

export default mergeSchemas({ schemas: [schema], resolvers });
