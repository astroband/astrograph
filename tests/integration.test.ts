import schema from "../src/schema";
import fs from "fs";
import { graphql } from "graphql";
import { Network } from "stellar-base";

Network.useTestNetwork();

const testCases = [
  "Single account query",
];

for (const testCaseName of testCases) {
  test(testCaseName, async () => {
    const queryFile = testCaseName.toLowerCase().replace(/ /g, "_");
    const query = fs.readFileSync(`${__dirname}/integration_queries/${queryFile}.gql`, "utf8");

    const { data } = await graphql(schema, query);

    expect(data).toMatchSnapshot();
  });
}
