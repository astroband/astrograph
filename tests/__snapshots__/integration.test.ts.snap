// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single account query 1`] = `
Object {
  "account": Object {
    "data": Array [],
    "flags": Object {
      "authImmutable": false,
      "authRequired": false,
      "authRevokable": false,
    },
    "id": "GBDQLZ7UYKZLLEBYFZIHW6ARJJS6GHZXSGD62A3MNGAEX5OOQMFD5GZT",
    "inflationDest": null,
    "ledger": Object {
      "header": Object {
        "ledgerVersion": 10,
      },
      "seq": 11216547,
    },
    "sequenceNumber": "48174702539046912",
    "signerFor": Array [
      Object {
        "id": "GBDQLZ7UYKZLLEBYFZIHW6ARJJS6GHZXSGD62A3MNGAEX5OOQMFD5GZT",
        "trustLines": Array [
          Object {
            "asset": Object {
              "code": "XLM",
              "issuer": "GBRPYHIL2CI3FNQ4BXLFMNDLFJUNPU2HY3ZMFSHONUCEOASW7QC7OX2H",
              "native": true,
            },
          },
        ],
      },
    ],
    "signers": Array [
      Object {
        "signer": Object {
          "id": "GBDQLZ7UYKZLLEBYFZIHW6ARJJS6GHZXSGD62A3MNGAEX5OOQMFD5GZT",
        },
      },
    ],
    "thresholds": Object {
      "high": 0,
      "low": 0,
      "masterWeight": 1,
      "medium": 0,
    },
    "trustLines": Array [
      Object {
        "account": Object {
          "id": "GBDQLZ7UYKZLLEBYFZIHW6ARJJS6GHZXSGD62A3MNGAEX5OOQMFD5GZT",
        },
        "asset": Object {
          "code": "XLM",
          "issuer": "GBRPYHIL2CI3FNQ4BXLFMNDLFJUNPU2HY3ZMFSHONUCEOASW7QC7OX2H",
          "native": true,
        },
      },
    ],
  },
}
`;
